File: Src/Icm.Core/Functions/TotalOrder/DoubleTotalOrder.vb
Namespace Icm
    Public Class DoubleTotalOrder
        Inherits BaseTotalOrder(Of Double)

        Public Overrides Function Least() As Double
            Return Double.NegativeInfinity
        End Function

        Public Overrides Function Greatest() As Double
            Return Double.PositiveInfinity
        End Function

        Public Overrides Function T2Long(ByVal t As Double) As Long
            Return BitConverter.DoubleToInt64Bits(t)
        End Function

        Public Overrides Function Long2T(ByVal d As Long) As Double
            Return BitConverter.Int64BitsToDouble(d)
        End Function

        Public Overrides Function [Next](ByVal t As Double) As Double
            Return Extreme.FloatingPoint.NextAfter(t, Double.MaxValue)
        End Function

    End Class
End Namespace

File: Src/Icm.Core/Functions/TotalOrder/DoubleTotalOrder.cs
using System;

namespace Icm
{
	public class DoubleTotalOrder : BaseTotalOrder<double>
	{

		public override double Least()
		{
			return double.NegativeInfinity;
		}

		public override double Greatest()
		{
			return double.PositiveInfinity;
		}

		public override long T2Long(double t)
		{
			return BitConverter.DoubleToInt64Bits(t);
		}

		public override double Long2T(long d)
		{
			return BitConverter.Int64BitsToDouble(d);
		}

		public override double Next(double t)
		{
			return Extreme.FloatingPoint.FloatingPointModule.NextAfter(t, double.MaxValue);
		}

	}
}